### ISSUE TITLE
Add support for on-demand metadata/lock table

### ISSUE THREAD

**Issue by acm19** [spring-cloud/spring-cloud-stream-binder-aws-kinesis#145]
When there is a lot of Kinesis re-sharding. Even with configuration like the following:
```
checkpoint-mode: periodic
checkpoint-interval: 60000
shard-iterator-type: LATEST
listener-mode: batch
```
The amount of reads in the metadata table increases exponentially. And the following error materialises:
```
com.amazonaws.services.dynamodbv2.model.ProvisionedThroughputExceededException: The level of configured provisioned throughput for the table was exceeded. Consider increasing your provisioning level with the UpdateTable API. (Service: AmazonDynamoDBv2; Status Code: 400; Error Code: ProvisionedThroughputExceededException; Request ID: XXXXXXX)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1695)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1350)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1101)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:758)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:732)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:714)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:674)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:656)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:520)
	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.doInvoke(AmazonDynamoDBClient.java:4192)
	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.invoke(AmazonDynamoDBClient.java:4159)
	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.executeGetItem(AmazonDynamoDBClient.java:2028)
	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.getItem(AmazonDynamoDBClient.java:1995)
	at com.amazonaws.services.dynamodbv2.document.internal.GetItemImpl.doLoadItem(GetItemImpl.java:77)
	at com.amazonaws.services.dynamodbv2.document.internal.GetItemImpl.getItemOutcome(GetItemImpl.java:40)
	at com.amazonaws.services.dynamodbv2.document.internal.GetItemImpl.getItemOutcome(GetItemImpl.java:99)
	at com.amazonaws.services.dynamodbv2.document.internal.GetItemImpl.getItem(GetItemImpl.java:111)
	at com.amazonaws.services.dynamodbv2.document.Table.getItem(Table.java:624)
	at org.springframework.integration.aws.metadata.DynamoDbMetadataStore.get(DynamoDbMetadataStore.java:263)
	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter.detectShardsToConsume(KinesisMessageDrivenChannelAdapter.java:630)
	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter.detectShardsToConsume(KinesisMessageDrivenChannelAdapter.java:617)
	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter.lambda$populateShardsForStream$0(KinesisMessageDrivenChannelAdapter.java:682)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
That's preventing the application from reading from a Kinesis Stream. Given stream resharding happen often we need to have more capacity in this table. It would be nice to be able to configure these tables as on-demand and forget about this issue.

**Comment by artembilan** [comment]
See if this metadata store options help you: https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/blob/master/spring-cloud-stream-binder-kinesis-docs/src/main/asciidoc/overview.adoc#metadatastore.

On the other hand it is already out-of this project scope and I think you are free to use a `DynamoDbClient` API directly in your application to adjust those options at runtime. Probably when you caught that `ProvisionedThroughputExceededException` in the error handler: https://docs.spring.io/spring-cloud-stream/docs/3.0.8.RELEASE/reference/html/spring-cloud-stream.html#spring-cloud-stream-overview-error-handling

**Comment by acm19** [comment]
Hi @artembilan,

I solved the issue by changing those tables to [on-demand](https://aws.amazon.com/blogs/aws/amazon-dynamodb-on-demand-no-capacity-planning-and-pay-per-request-pricing/) instead of provisioned capacity. That option is not currently available as part of the metadata store options. It's worth mentioning I had the same problem with the lock table but with the write capacity instead.

I don't think the second approach is a good idea because of the limitations of Increasing of [Decreasing Throughput (for Provisioned Tables) ](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html). Also, because AWS supports that out of the box.

Basically, what I'm asking is to add the option in the metadata store to create those table as [on-demand](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadWriteCapacityMode.html#HowItWorks.OnDemand) to begin with. It is possible create the tables manually beforehand but given it offers the possibility to create those tables, it would be nice to create a production ready table. We are actually using the feature when deploying to different envs and it would be very helpful.

I have a very busy week ahead but I can probably do a contribution if this feature makes sense to you.

**Comment by artembilan** [comment]
Thank you for pointing to AWS docs.
I didn't know about this billing mode feature. 😄 

Since we are talking about metadata store and lock registry tables, we need to make changes in the https://github.com/spring-projects/spring-integration-aws/ and then update this one exposing respective `BillingMode` configuration property: https://github.com/spring-projects/spring-integration-aws/issues/183

I'm looking forward to review your contribution!

**Comment by acm19** [comment]
Hi @artembilan, I was a bit busy and failed to raise the PR until now. I've created both PRs but this one is failing as we'll probably need to merge the other one first. Let me know if you want anything changed.

**Comment by acm19** [comment]
I've updated this PR as well @artembilan.
